src/rules/no_undef.rs

```diff
@@ -0,0 +1,345 @@
+  fn new() -> Box<Self> {
+    Box::new(NoUndef)
+  }
+
+  fn code(&self) -> &'static str {
+    "no-undef"
+  }
+
+  fn lint_module(&self, context: Arc<Context>, module: &Module) {
+    let mut collector = TopLevelBindingCollector {
+      top_level_ctxt: context.top_level_ctxt,
+      declared: Default::default(),
+    };
+    module.visit_with(module, &mut collector);
+
+    let mut visitor = NoGlobalAssignVisitor::new(context, collector.declared);
+    module.visit_with(module, &mut visitor);
+  }
+}
+
+/// Collects top level bindings, which have top level syntax context passed to the resolver.
+struct TopLevelBindingCollector {
+  /// The syntax context of the top level binding.
+  ///
+  /// Top level bindings and unresolved reference identifiers are marked with this.
+  top_level_ctxt: SyntaxContext,
+  /// If there exists a binding with such id, it's not global.
+  declared: HashSet<Id>,
+}
+
+impl TopLevelBindingCollector {
+  fn declare(&mut self, i: Id) {
+    // Optimization
+    if i.1 != self.top_level_ctxt {
+      return;
+    }
+    self.declared.insert(i);
+  }
+}
+
+impl Visit for TopLevelBindingCollector {
+  noop_visit_type!();
+
+  fn visit_fn_decl(&mut self, f: &FnDecl, _: &dyn Node) {
+    self.declare(f.ident.to_id());
+  }
+  fn visit_class_decl(&mut self, f: &ClassDecl, _: &dyn Node) {
+    self.declare(f.ident.to_id());
+  }
+
+  fn visit_import_named_specifier(
+    &mut self,
+    i: &ImportNamedSpecifier,
+    _: &dyn Node,
+  ) {
+    self.declare(i.local.to_id());
+  }
+
+  fn visit_import_default_specifier(
+    &mut self,
+    i: &ImportDefaultSpecifier,
+    _: &dyn Node,
+  ) {
+    self.declare(i.local.to_id());
+  }
+
+  fn visit_import_star_as_specifier(
+    &mut self,
+    i: &ImportStarAsSpecifier,
+    _: &dyn Node,
+  ) {
+    self.declare(i.local.to_id());
+  }
+
+  fn visit_var_declarator(&mut self, v: &VarDeclarator, _: &dyn Node) {
+    let ids: Vec<Id> = find_ids(&v.name);
+    for id in ids {
+      self.declare(id);
+    }
+  }
+}
+
+struct NoGlobalAssignVisitor {
+  context: Arc<Context>,
+  declared: HashSet<Id>,
+}
+
+impl NoGlobalAssignVisitor {
+  fn new(context: Arc<Context>, declared: HashSet<Id>) -> Self {
+    Self { context, declared }
+  }
+
+  fn check(&self, ident: &Ident) {
+    // We don't care about local references
+    // Because of this if statement, we can check for Map in
+    //
+    // function foo(Map) { ... }
+    //
+    if ident.span.ctxt != self.context.top_level_ctxt {
+      return;
+    }
+
+    // Ignore top level bindings declared in the file.
+    if self.declared.contains(&ident.to_id()) {
+      return;
+    }
+
+    // Globals
+    match &*ident.sym {
+      "Object"
+      | "Array"
+      | "Number"
+      | "Boolean"
+      | "String"
+      | "Promise"
+      | "Date"
+      | "NaN"
+      | "isNaN"
+      | "hasOwnProperty"
+      | "eval"
+      | "toString"
+      | "WeakRef"
+      | "Map"
+      | "FinalizationRegistry" => return,
```

Comment:
@kdy1 could you make it a static slice at top level and add `Deno` as well?

Reply:
Yes, I'll do it soon.